packages/neon-api/src/transaction/validator.ts

```diff
@@ -0,0 +1,314 @@
+   * @param autoFix will automatically fix transaction if specified as true
+   */
+  public async validateValidUntilBlock(
+    autoFix = false
+  ): Promise<ValidationResult> {
+    const { validUntilBlock } = this.transaction;
+    const height = await this.rpcClient.getBlockCount();
+    if (
+      validUntilBlock <= height ||
+      validUntilBlock >= height + tx.Transaction.MAX_TRANSACTION_LIFESPAN
+    ) {
+      const suggestion = tx.Transaction.MAX_TRANSACTION_LIFESPAN + height - 1;
+      if (autoFix) {
+        this.transaction.validUntilBlock = suggestion;
+        return {
+          valid: true,
+          suggestions: {
+            validUntilBlock: {
+              fixed: true,
+              prev: validUntilBlock,
+              suggestion
+            }
+          }
+        };
+      }
+      return {
+        valid: false,
+        suggestions: {
+          validUntilBlock: {
+            fixed: false,
+            prev: validUntilBlock,
+            suggestion
+          }
+        }
+      };
+    }
+    return {
+      valid: true
+    };
+  }
+
+  private async _validateIntent(
+    intent: sc.ScriptIntent
+  ): Promise<string | null> {
+    const { scriptHash, operation } = intent;
+    const manifest = await this.rpcClient.getContractState(scriptHash);
+    if (manifest === null) {
+      return `Unknown contract ${scriptHash}`;
+    }
+
+    if (operation) {
+      if (
+        manifest.abi.methods.map(method => method.name).indexOf(operation) < 0
+      ) {
+        return `Unknown method ${operation} for contract ${scriptHash}`;
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * // TODO: validate script oneline by invokeScript
+   * Validate intents
+   */
+  public async validateScript(): Promise<ValidationResult> {
+    const intents = new sc.ScriptParser(
+      this.transaction.script
+    ).toScriptParams();
+    const intentsRes: Array<string> = [];
+    await Promise.all(
+      intents.map(intent =>
+        this._validateIntent(intent).then(res =>
+          res ? intentsRes.push(res) : res
+        )
+      )
+    );
+    if (intentsRes.length > 0) {
+      return {
+        valid: false,
+        suggestions: {
+          script: {
+            message: intentsRes.join(";")
+          }
+        }
+      };
+    } else {
+      return {
+        valid: true
+      };
+    }
+  }
+
+  /**
+   * validate systemFee
+   * @param autoFix will automatically fix transaction if specified as true
+   */
+  public async validateSystemFee(autoFix = false): Promise<ValidationResult> {
+    const validateIntentsResult = await this.validateScript();
+    if (!validateIntentsResult.valid) {
+      return validateIntentsResult;
+    }
+
+    const { script, systemFee } = this.transaction;
+    const { gas_consumed: gasConsumed } = await this.rpcClient.invokeScript(
+      script
+    );
+    const minimumSystemFee = new u.Fixed8(
+      parseFloat(gasConsumed) * 1e-8
```

Comment:
please dont do multiply outside of Fixed8.

Reply:
Why? If I use `multiplyBy` it will return `BigNumber` and cannot use `ceil` after that.

Reply:
doing multiplication in js introduces inaccuracies. There is `times` or `mul`. have you tried that?

Reply:
OK